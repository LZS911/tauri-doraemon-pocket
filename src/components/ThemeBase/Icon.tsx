import { Badge, Tooltip } from 'antd';
import classNames from 'classnames';
import React from 'react';
import { IIconProps } from '.';
import useTheme, {
  getCurrentColorSchemeStrings,
} from '../../customHooks/useTheme';

const Icon: React.FC<IIconProps> = ({
  icon,
  className = '',
  hidden = false,
  isHoverCls = true,
  isWaterWaveCls = true,
  badge,
  tooltip,
  ...props
}) => {
  const { isDark, currentColorScheme } = useTheme();

  const genBadge = (el: React.ReactNode) => {
    return badge ? (
      <Badge
        {...badge}
        size="small"
        offset={[0, 2]}
        color={getCurrentColorSchemeStrings(isDark, currentColorScheme)[0]}
        className="leading-[0.875rem]"
      >
        <span className="dark:text-white">{el}</span>
      </Badge>
    ) : (
      el
    );
  };

  const genToolTip = (el: React.ReactNode) => {
    return tooltip ? <Tooltip {...tooltip}>{el}</Tooltip> : <>{el}</>;
  };

  return genToolTip(
    <div
      {...props}
      hidden={hidden}
      className={classNames(
        {
          ['hover:bg-slate-200 dark:hover:bg-slate-400']: isHoverCls,
        },
        {
          ['after:water-wave-hide active:after:water-wave-show']:
            isWaterWaveCls,
        },
        className,
        'transparent relative flex cursor-pointer items-center justify-center rounded-md bg-transparent p-2 transition-colors dark:text-white'
      )}
    >
      {genBadge(icon)}
    </div>
  );
};

export default Icon;
