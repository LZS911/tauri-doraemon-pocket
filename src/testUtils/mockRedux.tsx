import { configureStore } from '@reduxjs/toolkit';
import { combineReducers } from 'redux';
import layout from '../store/layout';
import config from '../store/config';
import { Dictionary } from '../typing/common.type';
import { Provider } from 'react-redux';
import React, { ReactNode } from 'react';
const reducers = combineReducers({
  layout,
  config,
});
export const storeFactory: any = (initStore: Dictionary = {}) => {
  return configureStore({
    reducer: reducers,
    preloadedState: initStore,
  });
};

export const CustomProvider: React.FC<{
  initStore: Dictionary;
  children: ReactNode;
}> = (props) => {
  return (
    <Provider store={storeFactory(props.initStore)}>{props.children}</Provider>
  );
};
