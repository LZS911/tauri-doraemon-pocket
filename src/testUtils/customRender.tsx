/* eslint-disable @typescript-eslint/no-explicit-any */
import { render, renderHook } from '@testing-library/react';
import { Provider } from 'react-redux';
import {
  BrowserRouter,
  MemoryRouter,
  MemoryRouterProps,
  Router,
  RouterProps,
} from 'react-router-dom';
import { Dictionary } from '../typing/common.type';
import { storeFactory } from './mockRedux';
type RenderParams = Parameters<typeof render>;

export const renderWithRedux: (arg: any) => ReturnType<typeof render> = (
  ...[ui, option, initStore]: [...RenderParams, Dictionary?]
) => {
  return render(
    <Provider store={storeFactory(initStore)}>{ui}</Provider>,
    option
  );
};
export const renderWithRouter: any = (...[ui, option]: [...RenderParams]) => {
  return render(<BrowserRouter>{ui}</BrowserRouter>, option);
};

export const renderHooksWithRedux = <TProps, TResult>(
  hooks: (props: TProps) => TResult,
  storeState: Dictionary = {}
) => {
  return renderHook(hooks, {
    wrapper: Provider,
    initialProps: {
      store: storeFactory(storeState),
    },
  } as any);
};

export const renderWithServerRouter: any = (
  ...[ui, option, props]: [...RenderParams, RouterProps]
) => {
  return render(<Router {...props}>{ui}</Router>, option);
};

export const renderWithServerRouterAndRedux: any = (
  ...[ui, option, props]: [...RenderParams, RouterProps]
) => {
  return render(
    <Router {...props}>
      {' '}
      <Provider store={storeFactory()}>{ui}</Provider>,
    </Router>,
    option
  );
};

export const renderWithMemoryRouter: any = (
  ...[ui, option, props]: [...RenderParams, MemoryRouterProps?]
) => {
  return render(<MemoryRouter {...props}>{ui}</MemoryRouter>, option);
};
