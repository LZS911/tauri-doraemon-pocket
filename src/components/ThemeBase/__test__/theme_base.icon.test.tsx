import Icon from '../Icon';
import { renderWithRedux } from '../../../testUtils/customRender';
import { fireEvent } from '@testing-library/react';

const icon = <>icon</>;

describe('test ThemeBase.Icon', () => {
  test('should match snapshot', () => {
    const { container } = renderWithRedux(<Icon icon={icon} />);
    expect(container).toMatchSnapshot();
  });

  test('should be as expected when isHoverCls is equal false', () => {
    const { container } = renderWithRedux(
      <Icon icon={icon} isHoverCls={false} />
    );
    expect(container).toMatchSnapshot();
  });

  test('should be as expected when isWaterWaveCls is equal false', () => {
    const { container } = renderWithRedux(
      <Icon icon={icon} isWaterWaveCls={false} />
    );
    expect(container).toMatchSnapshot();
  });

  test('should hide icon when hidden is equal true', () => {
    const { container } = renderWithRedux(<Icon icon={icon} hidden={true} />);
    // expect(container.querySelector('div')).toHaveAttribute('hidden');
  });

  test('should trigger click event when clicking icon', () => {
    const onClickSpy = jest.fn();
    const { container } = renderWithRedux(
      <Icon icon={icon} onClick={onClickSpy} />
    );
    expect(onClickSpy).toBeCalledTimes(0);
    fireEvent.click(container.querySelector('div')!);
    expect(onClickSpy).toBeCalledTimes(1);
  });

  test('should match snapshot when badge is not empty', () => {
    const { container } = renderWithRedux(
      <Icon icon={icon} badge={{ count: 2 }} />
    );
    expect(container).toMatchSnapshot();
  });

  test('should match snapshot when tooltip is not empty', () => {
    const { baseElement } = renderWithRedux(
      <Icon icon={icon} tooltip={{ title: 'title' }} />
    );
    expect(baseElement).toMatchSnapshot();
  });
});
