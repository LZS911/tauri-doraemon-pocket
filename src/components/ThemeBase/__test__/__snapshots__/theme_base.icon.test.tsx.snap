// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test ThemeBase.Icon should be as expected when isHoverCls is equal false 1`] = `
<div>
  <div
    class="after:water-wave-hide active:after:water-wave-show transparent relative flex cursor-pointer items-center justify-center rounded-md bg-transparent p-2 transition-colors dark:text-white"
  >
    icon
  </div>
</div>
`;

exports[`test ThemeBase.Icon should be as expected when isWaterWaveCls is equal false 1`] = `
<div>
  <div
    class="hover:bg-slate-200 dark:hover:bg-slate-400 transparent relative flex cursor-pointer items-center justify-center rounded-md bg-transparent p-2 transition-colors dark:text-white"
  >
    icon
  </div>
</div>
`;

exports[`test ThemeBase.Icon should match snapshot 1`] = `
<div>
  <div
    class="hover:bg-slate-200 dark:hover:bg-slate-400 after:water-wave-hide active:after:water-wave-show transparent relative flex cursor-pointer items-center justify-center rounded-md bg-transparent p-2 transition-colors dark:text-white"
  >
    icon
  </div>
</div>
`;

exports[`test ThemeBase.Icon should match snapshot when badge is not empty 1`] = `
<div>
  <div
    class="hover:bg-slate-200 dark:hover:bg-slate-400 after:water-wave-hide active:after:water-wave-show transparent relative flex cursor-pointer items-center justify-center rounded-md bg-transparent p-2 transition-colors dark:text-white"
  >
    <span
      class="ant-badge ant-badge-status leading-[0.875rem]"
    >
      <span
        class="dark:text-white"
      >
        icon
      </span>
      <sup
        class="ant-scroll-number ant-badge-count ant-badge-count-sm"
        data-show="true"
        style="margin-top: 2px; right: 0px; background: rgb(24, 144, 255);"
        title="2"
      >
        <span
          class="ant-scroll-number-only"
          style="transition: none;"
        >
          <span
            class="ant-scroll-number-only-unit current"
          >
            2
          </span>
        </span>
      </sup>
    </span>
  </div>
</div>
`;

exports[`test ThemeBase.Icon should match snapshot when tooltip is not empty 1`] = `
<body>
  <div>
    <div
      class="hover:bg-slate-200 dark:hover:bg-slate-400 after:water-wave-hide active:after:water-wave-show transparent relative flex cursor-pointer items-center justify-center rounded-md bg-transparent p-2 transition-colors dark:text-white"
    >
      icon
    </div>
  </div>
</body>
`;
